hacspec/src/machine_integers.rs

```diff
@@ -138,157 +223,258 @@ macro_rules! implement_public_mi {
             fn max_val() -> Self {
                 Self::from(<$base>::max_value())
             }
 
+            #[cfg_attr(feature="use_attributes", library(hacspec))]
             fn wrap_add(self, rhs: Self) -> Self {
                 self + rhs
             }
+            #[cfg_attr(feature="use_attributes", library(hacspec))]
             fn wrap_sub(self, rhs: Self) -> Self {
                 self - rhs
             }
+            #[cfg_attr(feature="use_attributes", library(hacspec))]
             fn wrap_mul(self, rhs: Self) -> Self {
                 self * rhs
             }
+            #[cfg_attr(feature="use_attributes", library(hacspec))]
             fn wrap_div(self, _rhs: Self) -> Self {
                 unimplemented!();
             }
 
             /// `self ^ exp` where `exp` is a `u32`.
+            /// **Note:** the exponent `exp` MUST not be secret.
+            #[cfg_attr(feature="use_attributes", library(hacspec))]
             fn pow(self, exp: u32) -> Self {
-                let s = <$t>::declassify(self);
-                Self::from(s.pow(exp))
+                let mut s = self;
+                if exp == 0 {
+                    return <$t>::from(1 as $base);
+                } else {
+                    for _ in 1..exp {
+                        s = s * self
+                    }
+                }
+                Self::from(s)
             }
             /// `self ^ exp` where `exp` is a `Self`.
+            /// Here both, base and exponent, are secret.
+            #[cfg_attr(feature="use_attributes", library(hacspec))]
             fn pow_self(self, _exp: Self) -> Self {
                 unimplemented!();
             }
-            /// (self - rhs) % n.
-            fn sub_mod(self, rhs: Self, n: Self) -> Self {
-                let s = <$t>::declassify(self);
-                let o = <$t>::declassify(rhs);
-                let n = <$t>::declassify(n);
-                Self::from((s - o) % n)
-            }
-            /// `(self + rhs) % n`
-            fn add_mod(self, rhs: Self, n: Self) -> Self {
-                let s = <$t>::declassify(self);
-                let o = <$t>::declassify(rhs);
-                let n = <$t>::declassify(n);
-                Self::from((s + o) % n)
-            }
-            /// `(self * rhs) % n`
-            fn mul_mod(self, rhs: Self, n: Self) -> Self{
-                let s = <$t>::declassify(self);
-                let o = <$t>::declassify(rhs);
-                let n = <$t>::declassify(n);
-                Self::from((s * o) % n)
-            }
-            /// `(self ^ exp) % n`
-            fn pow_mod(self, _exp: Self, _n: Self) -> Self {
-                unimplemented!();
-            }
             /// Division.
+            #[cfg_attr(feature="use_attributes", library(hacspec))]
             fn div(self, rhs: Self) -> Self {
                 let s = <$t>::declassify(self);
                 let o = <$t>::declassify(rhs);
```

Comment:
So here, you took the option of declassifying silently. My opinion is that the code should `panic!` here, because we strictly disallow non-constat-time operations on secret integers.

Reply:
This again should use `ct_div`.

Reply:
Ok !